### ISSUE TITLE
Question about Reconstruction ability

### ISSUE THREAD

**Issue by zhentao-zou** [selftok-team/SelftokTokenizer#7]
Thanks for a nice work! I have reproduced the reconstructions result in imageNet Validation 50K according to your yaml and pretrained models. However, the results is 2dB lower than the psnr results the paper reported 

![Image](https://github.com/user-attachments/assets/45525bc0-d09b-400b-baed-1619f3fa5cc7)
![Image](https://github.com/user-attachments/assets/09a27ca5-8ed5-415b-b9d8-1d7fff326939)

The following is the test code

AVE_DIR_Ori = pathlib.Path("/home/ma-user/work/z50048519/Selftok/ori_imgs")
SAVE_DIR_Ori.mkdir(parents=True, exist_ok=True)

INC_MEAN = torch.tensor([0.485, 0.456, 0.406])
INC_STD  = torch.tensor([0.229, 0.224, 0.225])
preproc_inc = transforms.Compose([
    transforms.Resize(299, antialias=True),
    transforms.CenterCrop(299),
    transforms.ToTensor(),
    transforms.Normalize(INC_MEAN, INC_STD),
])

from torchvision.models import inception_v3

inception = inception_v3(weights=None, aux_logits=True)  
state = torch.load("/home/ma-user/work/z50048519/pretrained/inception_v3_google-0cc3c7bd.pth",
                   map_location=device)
inception.load_state_dict(state, strict=True)

inception.fc = nn.Identity() 

inception.eval().to(device)

lpips_net = lpips.LPIPS(net="alex").eval().to(device)

preproc_01 = transforms.Compose([
    transforms.Resize(256, antialias=True),
    transforms.CenterCrop(256),
    transforms.ToTensor(),              # → [0,1]
])

#import os

import sys
print(sys.path)
sys.path.append(".")

import argparse
from mimogpt.infer.infer_utils import parse_args_from_yaml
from torchvision import transforms
from PIL import Image
import torch
import numpy as np
from mimogpt.infer.SelftokPipeline import SelftokPipeline
from mimogpt.infer.SelftokPipeline import NormalizeToTensor
from torchvision.utils import save_image
from pathlib import Path
parser = argparse.ArgumentParser()
#parser.add_argument("--yml-path", type=str, default="/home/ma-user/work/z50048519/Selftok/configs/res256/256-eval.yml") # download from https://huggingface.co/stabilityai/stable-diffusion-3-medium/resolve/main/sd3_medium.safetensors?download=true, require huggingface login, you have to change the format to .pt with safetensor_to_pt.py
parser.add_argument("--yml-path", type=str, default="/home/ma-user/work/z50048519/Selftok/configs/renderer/renderer-eval.yml")
#parser.add_argument("--pretrained", type=str, default="/home/ma-user/work/z50048519/Selftok/weight/tokenizer_1024_ckpt.pth") 
parser.add_argument("--pretrained", type=str, default="/home/ma-user/work/z50048519/Selftok/weight/renderer_512_ckpt.pth")
parser.add_argument("--sd3_pretrained", type=str, default="/home/ma-user/work/z50048519/Selftok/SD3") 
parser.add_argument("--data_size", type=int, default=256)

args = parser.parse_args()

cfg = parse_args_from_yaml(args.yml_path)
model = SelftokPipeline(cfg=cfg, ckpt_path=args.pretrained, sd3_path=args.sd3_pretrained, datasize=args.data_size, device='cuda')

img_transform = transforms.Compose([
    transforms.Resize(args.data_size),
    transforms.CenterCrop(args.data_size),
    NormalizeToTensor(),
])

def denorm_01(x):
    return x.mul(0.5).add(0.5).clamp_(0, 1)

def norm11(x):        # [0,1]  → [-1,1]
    return x * 2.0 - 1.0

 count = 0
 for file in files:
     pf = pq.ParquetFile(file)
     for batch in pf.iter_batches(batch_size=1000):
         images_ref = batch.column("image").to_pylist()      # 原图
         #img_ref = Image.open(io.BytesIO(images_ref[0]))
       
        
         ref_bytes_list = [d["bytes"] for d in images_ref]  # TODO: 换成真解码结果
         images=[img_transform(Image.open(io.BytesIO(b))) for b in ref_bytes_list]
         images = torch.stack(images).to('cuda')
         for b in range(images.size(0)):
             src_t = images[b] 
             src_t =src_t*0.5+0.5
             to_pil_image(src_t.to(torch.float32).clamp(0, 1).cpu()).save(SAVE_DIR_Ori / f"ori_{count:05d}.jpg")
             count = count+1

SAVE_DIR_Ori = Path("/home/ma-user/work/z50048519/Selftok/ori_imgs")
image_paths = sorted(SAVE_DIR_Ori.glob("*.jpg"))

count = 0
batch_size = 250
for i in range(0, len(image_paths), batch_size):
    batch_paths = image_paths[i : i + batch_size]
    batch = [img_transform(Image.open(p)) for p in batch_paths]
    batch = torch.stack(batch).to("cuda")   # (B, C, H, W)
        
    with torch.no_grad():
        tokens = model.encoding(batch, device='cuda')
        #np.save('./token.npy', tokens.detach().cpu().numpy())

        #tokens = np.load('./token.npy')
        rec_images =model.decoding_with_renderer(tokens, device='cuda')

    for b in range(batch.size(0)):
        src_t = batch[b]         
        rec_t = rec_images[b]
            #src_t = (src_t+1)/2
        src_t = denorm_01(src_t)

   

        to_pil_image(rec_t.to(torch.float32).clamp(0, 1).cpu()).save(SAVE_DIR / f"dec_{count:05d}.jpg")
 

       
        src_ = src_t.unsqueeze(0)          
        rec_ = rec_t.unsqueeze(0)

        psnr_list.append(psnr(rec_, src_, data_range=1.0).item())
        ssim_list.append(ssim(rec_, src_, data_range=1.0).item())

        with torch.no_grad():
            lpips_val = lpips_net(rec_ * 2 - 1, src_ * 2 - 1).item()  # ↗ 对 [-1,1]
        lpips_list.append(lpips_val)

**Comment by SainingZhang** [comment]
Thanks for trying Selftok and for sharing your results! 
We noticed that testing PSNR using different methods can lead to some discrepancies. Could you please refer to the code below, rerun the test, and then share your feedback with us?

```python
import numpy as np
from torchvision.utils import save_image

def compute_psnr(recon, ori):
  x = torch.from_numpy(np.array(Image.open(ori))).cuda().float()
  y = torch.from_numpy(np.array(Image.open(recon))).cuda().float()
  mse = F.mse_loss(y, x)
  psnr = 20 * torch.log10(torch.Tensor([255.0]).to(ori.device)) - 10 * torch.log10(mse)

  return psnr

def norm_ip(img, low, high):
    img.clamp_(min=low, max=high)
    img.sub_(low).div_(max(high - low, 1e-5))

norm_ip(batch, -1, 1)
for b in range(len(rec_images)):
    save_image(rec_images[b], f"recon_{b}.png")
    save_image(batch[b], f"ori_{b}.png")
    psnr = compute_psnr(f"recon_{b}.png", f"ori_{b}.png")
```
